### DISCUSSION TITLE
SHA-1 in trust chain validation during peer certificate authentication

### DISCUSSION THREAD

**Discussion by amdrsantos** [strongswan/strongswan#1577] [Q&A] [answered]
Is SHA-1 still possible to use in the trust chain validation during peer certificate authentication?

If yes, is there any way that it can be disabled the use of SHA-1 in trust chain validation?

**Comment by tobiasbrunner** [comment, marked as answer]
> Is SHA-1 still possible to use in the trust chain validation during peer certificate authentication?

Yes, strongSwan does currently not restrict this by default.



> If yes, is there any way that it can be disabled the use of SHA-1 in trust chain validation?

You can enforce certain signature algorithms via auth restrictions. For instance, to allow any public key authentication with SHA-2 only configure the following in `remote.auth`: `pubkey-sha256-sha384-sha512`. However, note that to allow legacy IKEv2 authentication with e.g. Windows clients, you either have to disable `charon.signature_authentication_ constraints` or add IKE constraints explicitly (e.g. `...-ike:pubkey-sha256-sha384-sha512-sha1`).

**Reply by amdrsantos** [reply]
I found some discrepancies while trying to use this functionality.
I am using:
 - The same root certificate with an SHA-1 hash signature on both ends
 - Both leaf certificates with SHA256 hash signature, signed with the root certificate from the point above

When I set one remote auth to:
```
        remote {
            auth = ike:pubkey-sha256-sha384-sha512
            id = GX2ALEX.mycompany.com
        }
```

The connection is still established.
I was expecting it to fail since there is a signing hash SHA-1 in the trust chain validation of the remote.

**Reply by tobiasbrunner** [reply]
The `ike:` prefix restricts the signatures used for IKE, not the signatures in the trust chain.

**Reply by amdrsantos** [reply]
Thanks, Tobias, 
Tested again with:
```
        remote {
            auth = pubkey-sha256-sha384-sha512
            id = GX2ALEX.mycompany.com
        }
```
I still have the connection established.

**Reply by tobiasbrunner** [reply]
I see. I missed the part about SHA-1 being used only for the self-signed root CA. That signature is currently ignored for these constraints. The root certificate has to be installed locally and trusted anyway, so the signature scheme isn't really an issue.

**Reply by amdrsantos** [reply]
Still, a minor doubt to do not allowing the SHA1 in the SIGNATURE_HASH_ALGORITHMS of the IKE_SA_INIT request, `remote.auth = pubkey-sha256-sha384-sha512` is enough?

**Reply by tobiasbrunner** [reply]
If `charon.signature_authentication_ constraints` is enabled (the default), then yes. Otherwise, you have to explicitly add constraints with `ike:` prefix.

**Comment by amdrsantos** [comment]
Hi Tobias,
I disabled the charon.signature_authentication_constraints.
But I am struggling to configure strongSwan such that:
* SHA-1 hash signature algorithm allowed in trusted-chain certificate verification
* SHA-2 is not allowed in the IKEv2 Signature Authentication

Can you post an example?
Thanks, Alex

**Reply by tobiasbrunner** [reply]
Either of the following should achieve that (instead of `pubkey` you could also use a specific algorithm like `rsa` or `ecdsa`, optionally followed by a minimum key size):

`remote.auth = ike:pubkey-sha256-sha384-sha512-pubkey-sha1-sha256-sha384-sha512`

`remote.auth = pubkey-sha1-sha256-sha384-sha512-ike:pubkey-sha256-sha384-sha512`

But as seen before, allowing SHA-1 for the trust chain validation is only necessary if an end-entity or intermediate CA certificate is signed with SHA-1, which shouldn't be the case anymore nowadays. SHA-1 is more likely necessary for IKEv2 as not all clients support RFC 7427 style signature authentication (e.g. Windows, which will always use SHA-1 for RSA signatures).

**Reply by amdrsantos** [reply]
How do you do the same for psk instead of pubkey?

**Reply by tobiasbrunner** [reply]
You don't, as the PSK authentication is based on the negotiated PRF. So to avoid SHA-1 there, make sure `prfsha1` (PRF_HMAC_SHA1(2)) isn't negotiated as PRF (i.e. don't include `sha1` in your IKE proposal, or, if you want to allow that for integrity protection, explicitly configure the allowed PRFs `prfsha256-prfsha384-prfsha512` so integrity algorithms aren't mapped to PRFs automatically).
